src/cmap/auth/mongodb_oidc/callback_lock_cache.ts

```diff
@@ -0,0 +1,135 @@
+import { MongoInvalidArgumentError } from '../../../error';
+import type { Connection } from '../../connection';
+import type { MongoCredentials } from '../mongo_credentials';
+import type {
+  IdPServerInfo,
+  IdPServerResponse,
+  OIDCCallbackContext,
+  OIDCRefreshFunction,
+  OIDCRequestFunction
+} from '../mongodb_oidc';
+
+/** Error message for when request callback is missing. */
+const REQUEST_CALLBACK_REQUIRED_ERROR =
+  'Auth mechanism property REQUEST_TOKEN_CALLBACK is required.';
+/* Counter for function "hashes".*/
+let FN_HASH_COUNTER = 0;
+/* No function present function */
+const NO_FUNCTION: OIDCRequestFunction = () => {
+  return Promise.resolve({ accessToken: 'test' });
+};
+/* The map of function hashes */
+const FN_HASHES = new WeakMap<OIDCRequestFunction | OIDCRefreshFunction, number>();
+/* Put the no function hash in the map. */
+FN_HASHES.set(NO_FUNCTION, FN_HASH_COUNTER);
+
+/**
+ * An entry of callbacks in the cache.
+ */
+interface CallbacksEntry {
+  requestCallback: OIDCRequestFunction;
+  refreshCallback?: OIDCRefreshFunction;
+  callbackHash: string;
+}
+
+/**
+ * A cache of request and refresh callbacks per server/user.
+ */
+export class CallbackLockCache {
```

Comment:
This cache is now responsible for the hashing of the request and refresh functions, which is also then used by the `TokenEntryCache`. It also is responsible for wrapping the callbacks in a lock and storing them by the same cache key as the token results in the other cache. Main reasoning for not storing these in the other cache and creating a new one is that the callback functions have no expiration.